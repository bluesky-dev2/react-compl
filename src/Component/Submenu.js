import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import cssClasses from './../cssClasses';

export default class Submenu extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    title: PropTypes.string.isRequired,
    leftIcon: PropTypes.node,
    rightIcon: PropTypes.node,
    disabled: PropTypes.bool,
  };

  static defaultProps = {
    leftIcon: '',
    rightIcon: '',
    disabled: false,
    onClick: () => {},
    targetNode: {},
    data: null,
    refsFromProvider: []
  };

  render() {
    const className = cx(cssClasses.ITEM, {
      [`${cssClasses.ITEM_DISABLED}`]: this.props.disabled
    });

    return (
      <div className={className} role="presentation">
        <div className={cssClasses.ITEM_DATA}>
          {this.props.leftIcon}
          {this.props.title}
          {this.props.rightIcon}
          <span>â–¶</span>
        </div>
        <div className={cssClasses.SUBMENU}>{this.props.children}</div>
      </div>
    );
  }
}
