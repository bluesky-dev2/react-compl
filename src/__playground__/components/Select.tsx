import React, { ChangeEvent } from 'react';

interface SelectProps {
  name: string;
  value: string;
  data: string[];
  onChange: (e: ChangeEvent<HTMLSelectElement>) => void;
}

export default ({ name, value, data, onChange }: SelectProps) => (
  <select name={name} id={name} value={value} onChange={onChange}>
    {data.map(item => (
      <option key={item} value={item}>
        {item}
      </option>
    ))}
  </select>
);
