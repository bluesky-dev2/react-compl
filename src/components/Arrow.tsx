import React from 'react';

export const Arrow = () => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="18"
    height="18"
    viewBox="0 0 24 24"
    fill="none"
    stroke="currentColor"
    strokeWidth="2"
    strokeLinecap="round"
    strokeLinejoin="round"
  >
    <polyline points="9 18 15 12 9 6" />
  </svg>
);
