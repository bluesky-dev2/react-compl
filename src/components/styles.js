/*
* css classes mapping
* */
export default {
  menu: 'react-contexify',
  submenu: 'react-contexify react-contexify__submenu',
  submenuArrow: 'react-contexify__submenu-arrow',
  separator: 'react-contexify__separator',
  item: 'react-contexify__item',
  itemDisabled: 'react-contexify__item--disabled',
  itemContent: 'react-contexify__item__data',
  itemIcon: 'react-contexify__item__icon',
  theme: 'react-contexify__theme--',
  animationWillEnter: 'react-contexify__will-enter--',
  animationWillLeave: 'react-contexify__will-leave--'
};
