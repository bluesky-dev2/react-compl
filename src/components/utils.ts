import { Children, cloneElement, ReactNode, ReactElement } from 'react';

import { BooleanPredicate, PredicateParams, TriggerEvent } from '../types';

export function isFn(v: any): v is Function {
  return typeof v === 'function';
}

export function isStr(v: any): v is String {
  return typeof v === 'string';
}

export function cloneItems(
  children: ReactNode,
  props: { triggerEvent: TriggerEvent; propsFromTrigger?: object }
) {
  return Children.map(
    // remove null item
    Children.toArray(children).filter(Boolean),
    (item) => cloneElement(item as ReactElement<any>, props)
  );
}

export function getMousePosition(e: TriggerEvent) {
  const pos = {
    x: (e as MouseEvent).clientX,
    y: (e as MouseEvent).clientY,
  };

  const touch = (e as TouchEvent).changedTouches;

  if (touch) {
    pos.x = touch[0].clientX;
    pos.y = touch[0].clientY;
  }

  if (!pos.x || pos.x < 0) pos.x = 0;

  if (!pos.y || pos.y < 0) pos.y = 0;

  return pos;
}

export function getPredicateValue(
  predicate: BooleanPredicate,
  payload: PredicateParams
) {
  return isFn(predicate) ? predicate(payload) : predicate;
}
