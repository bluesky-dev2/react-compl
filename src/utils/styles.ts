/*
 * css classes mapping
 * */
export const styles = {
  menu: 'react-contexify',
  submenu: 'react-contexify react-contexify__submenu',
  submenuArrow: 'react-contexify__submenu-arrow',
  separator: 'react-contexify__separator',
  item: 'react-contexify__item',
  itemDisabled: 'react-contexify__item--disabled',
  itemContent: 'react-contexify__item__content',
  itemIcon: 'react-contexify__item__icon',
  theme: 'react-contexify__theme--',
  animationWillEnter: 'react-contexify__will-enter--',
  animationWillLeave: 'react-contexify__will-leave--'
};

// I could use enum instead for both interfaces

export interface BuiltInTheme {
  light: string;
  dark: string;
}

export interface BuiltInAnimation {
  fade: string;
  flip: string;
  pop: string;
  zoom: string;
}

export const theme: BuiltInTheme = {
  light: 'light',
  dark: 'dark'
};

export const animation: BuiltInAnimation = {
  fade: 'fade',
  flip: 'flip',
  pop: 'pop',
  zoom: 'zoom'
};
