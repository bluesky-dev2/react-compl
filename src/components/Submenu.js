import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './styles';

export default class Submenu extends Component {
  static propTypes = {
    label: PropTypes.node.isRequired,
    children: PropTypes.node.isRequired,
    arrow: PropTypes.node,
    nativeEvent: PropTypes.object,
    dataFromProvider: PropTypes.any,
    refsFromProvider: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.object)
    ]),
    className: PropTypes.string,
    style: PropTypes.object,
    disabled: PropTypes.oneOfType([PropTypes.bool, PropTypes.func])
  };

  static defaultProps = {
    arrow: 'â–¶',
    nativeEvent: null,
    dataFromProvider: null,
    refsFromProvider: null,
    className: null,
    style: {},
    disabled: false
  };

  state = {
    style: {
      left: '100%',
      top:0,
      bottom: 'initial'
    }
  };

  setRef = ref => {
    this.menu = ref;
  };

  componentDidMount() {
    const { innerWidth, innerHeight } = window;
    const rect = this.menu.getBoundingClientRect();
    const style = {};

    if(rect.right < innerWidth) {
      style.left = '100%';
    } else {
      style.right = '100%';
    }

    if (rect.bottom > innerHeight) {
      style.bottom = 0;
      style.top = 'initial';
    } else {
      style.bottom = 'initial';
      style.top = 0;
    }

    this.setState({
      style: style
    });
  }

  getMenuItem() {
    const { nativeEvent, refsFromProvider, dataFromProvider } = this.props;

    return React.Children.map(this.props.children, item =>
      React.cloneElement(item, {
        nativeEvent,
        refsFromProvider,
        dataFromProvider
      })
    );
  }

  handleClick(e) {
    e.stopPropagation();
  }

  render() {
    const {
      arrow,
      disabled,
      className,
      style,
      label,
      nativeEvent,
      refsFromProvider,
      dataFromProvider,
    } = this.props;
    const cssClasses = cx(styles.item, className, {
      [`${styles.itemDisabled}`]:
        typeof disabled === 'function'
          ? disabled({
              event: nativeEvent,
              dataFromProvider,
              ref: refsFromProvider
            })
          : disabled
    });
    const submenuStyle = {
      ...style,
      ...this.state.style
    };

    return (
      <div
        className={cssClasses}
        role="presentation"
      >
        <div className={styles.itemContent} onClick={this.handleClick}>
          {label}
          <span className={styles.submenuArrow}>{arrow}</span>
        </div>
        <div
          className={styles.submenu}
          ref={this.setRef}
          style={submenuStyle}
        >
          {this.getMenuItem()}
        </div>
      </div>
    );
  }
}
