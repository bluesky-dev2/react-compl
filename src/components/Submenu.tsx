import React, { Component, ReactNode, SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { cloneItem } from './cloneItem';
import { styles } from '../utils/styles';
import {
  MenuItemEventHandler,
  TriggerEvent,
  StyleProps,
  InternalProps
} from '../types';

export interface SubMenuProps extends StyleProps, InternalProps {
  /**
   * Any valid node that can be rendered
   */
  label: ReactNode;

  /**
   * Any valid node that can be rendered
   */
  children: ReactNode;

  /**
   * Render a custom arrow
   */
  arrow: ReactNode;

  /**
   * Disable or not the `Submenu`. If a function is used, a boolean must be returned
   */
  disabled: boolean | ((args: MenuItemEventHandler) => boolean);
}

interface SubMenuState {
  left?: string | number;
  right?: string | number;
  top?: string | number;
  bottom?: string | number;
}

class Submenu extends Component<SubMenuProps, SubMenuState> {
  static propTypes = {
    label: PropTypes.node.isRequired,
    children: PropTypes.node.isRequired,
    nativeEvent: PropTypes.object,
    arrow: PropTypes.node,
    disabled: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
    className: PropTypes.string,
    style: PropTypes.object
  };
  static defaultProps = {
    arrow: 'â–¶',
    disabled: false,
    nativeEvent: {} as TriggerEvent
  };

  state = {
    left: '100%',
    top: 0,
    bottom: 'initial'
  };

  menu!: HTMLElement;

  setRef = (ref: HTMLDivElement) => {
    this.menu = ref;
  };

  componentDidMount() {
    const { innerWidth, innerHeight } = window;
    const rect = this.menu.getBoundingClientRect();
    const style: SubMenuState = {};

    if (rect.right < innerWidth) {
      style.left = '100%';
      style.right = undefined;
    } else {
      style.right = '100%';
      style.left = undefined;
    }

    if (rect.bottom > innerHeight) {
      style.bottom = 0;
      style.top = 'initial';
    } else {
      style.bottom = 'initial';
      style.top = 0;
    }

    this.setState(style);
  }

  handleClick(e: SyntheticEvent) {
    e.stopPropagation();
  }

  render() {
    const {
      arrow,
      disabled,
      className,
      style,
      label,
      nativeEvent,
      children,
      propsFromTrigger
    } = this.props;

    const cssClasses = cx(styles.item, className, {
      [`${styles.itemDisabled}`]:
        typeof disabled === 'function'
          ? disabled({
              event: nativeEvent as TriggerEvent,
              props: { ...propsFromTrigger }
            })
          : disabled
    });

    const submenuStyle = {
      ...style,
      ...this.state
    };

    return (
      <div className={cssClasses} role="presentation">
        <div className={styles.itemContent} onClick={this.handleClick}>
          {label}
          <span className={styles.submenuArrow}>{arrow}</span>
        </div>
        <div className={styles.submenu} ref={this.setRef} style={submenuStyle}>
          {cloneItem(children, {
            propsFromTrigger,
            nativeEvent: nativeEvent as TriggerEvent
          })}
        </div>
      </div>
    );
  }
}

export { Submenu };
