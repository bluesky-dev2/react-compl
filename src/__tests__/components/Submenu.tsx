/* eslint-env jest */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Submenu } from '../../components/Submenu';
import { styles } from '../../utils/styles';

const Children = () => <div>foo</div>;
describe('Submenu', () => {
  it('Should render without crash 💥', () => {
    const component = mount(
      <Submenu label="bar">
        <Children />
      </Submenu>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should not go offscreen', () => {
    (window as any).innerWidth = 100;
    (window as any).innerHeight = 100;
    Element.prototype.getBoundingClientRect = () =>
      ({
        right: 200,
        bottom: 200
      } as ClientRect);
    const component = mount(
      <Submenu label="bar">
        <Children />
      </Submenu>
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('Can be disabled with true or a function', () => {
    let component = mount(
      <Submenu label="bar" disabled>
        <Children />
      </Submenu>
    );
    expect(component.html()).toMatch(styles.itemDisabled);

    component.unmount();
    component = mount(
      <Submenu label="bar" disabled={() => true}>
        <Children />
      </Submenu>
    );
    expect(component.html()).toMatch(styles.itemDisabled);
  });

  it('Should accept a custom arrow ➡️', () => {
    const component = mount(
      <Submenu label="bar" arrow="🦄">
        <Children />
      </Submenu>
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
