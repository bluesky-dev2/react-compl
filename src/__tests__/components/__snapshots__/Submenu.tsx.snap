// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submenu Should accept a custom arrow ‚û°Ô∏è 1`] = `
<Submenu
  arrow="ü¶Ñ"
  disabled={false}
  label="bar"
  nativeEvent={Object {}}
>
  <div
    className="react-contexify__item"
    role="presentation"
  >
    <div
      className="react-contexify__item__content"
      onClick={[Function]}
    >
      bar
      <span
        className="react-contexify__submenu-arrow"
      >
        ü¶Ñ
      </span>
    </div>
    <div
      className="react-contexify react-contexify__submenu"
      style={
        Object {
          "bottom": 0,
          "left": "100%",
          "right": "100%",
          "top": "initial",
        }
      }
    >
      <Children
        key=".$.0"
        nativeEvent={Object {}}
      >
        <div>
          foo
        </div>
      </Children>
    </div>
  </div>
</Submenu>
`;

exports[`Submenu Should not go offscreen 1`] = `
<Submenu
  arrow="‚ñ∂"
  disabled={false}
  label="bar"
  nativeEvent={Object {}}
>
  <div
    className="react-contexify__item"
    role="presentation"
  >
    <div
      className="react-contexify__item__content"
      onClick={[Function]}
    >
      bar
      <span
        className="react-contexify__submenu-arrow"
      >
        ‚ñ∂
      </span>
    </div>
    <div
      className="react-contexify react-contexify__submenu"
      style={
        Object {
          "bottom": 0,
          "left": "100%",
          "right": "100%",
          "top": "initial",
        }
      }
    >
      <Children
        key=".$.0"
        nativeEvent={Object {}}
      >
        <div>
          foo
        </div>
      </Children>
    </div>
  </div>
</Submenu>
`;

exports[`Submenu Should render without crash üí• 1`] = `
<Submenu
  arrow="‚ñ∂"
  disabled={false}
  label="bar"
  nativeEvent={Object {}}
>
  <div
    className="react-contexify__item"
    role="presentation"
  >
    <div
      className="react-contexify__item__content"
      onClick={[Function]}
    >
      bar
      <span
        className="react-contexify__submenu-arrow"
      >
        ‚ñ∂
      </span>
    </div>
    <div
      className="react-contexify react-contexify__submenu"
      style={
        Object {
          "bottom": "initial",
          "left": "100%",
          "top": 0,
        }
      }
    >
      <Children
        key=".$.0"
        nativeEvent={Object {}}
      >
        <div>
          foo
        </div>
      </Children>
    </div>
  </div>
</Submenu>
`;
